### PR TITLE
Add online inference notebook and cli examples for CLIP

### PR DESCRIPTION
**Pull request by bhimar**
# Description
Add online inference notebook and cli examples for CLIP (zero-shot-image-classification)

Demo endpoints:
[cli](https://ml.azure.com/endpoints/realtime/zero-shot-image-cls-1694188700/detail?wsid=/subscriptions/dbd697c3-ef40-488f-83e6-5ad4dfb78f9b/resourceGroups/rbhimani-rg/providers/Microsoft.MachineLearningServices/workspaces/rbhimani-ws&tid=72f988bf-86f1-41af-91ab-2d7cd011db47)
[notebook](https://ml.azure.com/endpoints/realtime/zero-shot-image-cls-1694727182/detail?wsid=/subscriptions/dbd697c3-ef40-488f-83e6-5ad4dfb78f9b/resourceGroups/rbhimani-rg/providers/Microsoft.MachineLearningServices/workspaces/rbhimani-ws&tid=72f988bf-86f1-41af-91ab-2d7cd011db47)

These endpoints were created with a model that I published to the registry rbhimani-registry because CLIP is not yet published in azureml registry.

# Checklist

- [ ] I have read the [contribution guidelines](https://github.com/Azure/azureml-examples/blob/main/CONTRIBUTING.md)
- [ ] I have coordinated with the docs team (mldocs@microsoft.com) if this PR deletes files or changes any file names or file extensions.
- [ ] Pull request includes test coverage for the included changes.

### CODE REVIEW

File: sdk/python/foundation-models/system/inference/zero-shot-image-classification/zero-shot-image-classification-online-endpoint.ipynb [hunk 1]
```diff
@@ -0,0 +1,384 @@
...
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "from azure.ai.ml.entities import OnlineRequestSettings, ProbeSettings\n",
+    "\n",
+    "deployment_name = \"zero-shot-image-mlflow-deploy\"\n",
+    "\n",
+    "print(foundation_model.id)\n",
+    "print(online_endpoint_name)\n",
+    "print(deployment_name)\n",
```
Comment (vivek-dani) [inline sdk/python/foundation-models/system/inference/zero-shot-image-classification/zero-shot-image-classification-online-endpoint.ipynb:229]: we can remove this, as we just initialized it above

File: sdk/python/foundation-models/system/inference/zero-shot-image-classification/zero-shot-image-classification-online-endpoint.ipynb [hunk 2]
```diff
@@ -0,0 +1,384 @@
...
+   "source": [
+    "demo_deployment = workspace_ml_client.online_deployments.get(\n",
+    "    name=deployment_name,\n",
+    "    endpoint_name=online_endpoint_name,\n",
+    ")\n",
+    "\n",
+    "# Get the details for online endpoint\n",
+    "endpoint = workspace_ml_client.online_endpoints.get(name=online_endpoint_name)\n",
+    "\n",
+    "# Existing traffic details\n",
```
Comment (vivek-dani) [inline sdk/python/foundation-models/system/inference/zero-shot-image-classification/zero-shot-image-classification-online-endpoint.ipynb:287]: We can remove this as we initialized to 100% just above. Even I removed it from stablediffusion notebook

**Review by vivek-dani — APPROVED** [review]

File: cli/foundation-models/system/inference/zero-shot-image-classification/prepare_data.py [hunk 3]
```diff
@@ -0,0 +1,111 @@
...
+import base64
+import json
+import os
+import shutil
+import urllib.request
+import pandas as pd
+from zipfile import ZipFile
+
+# Change this to match the inference dataset
+LABELS = "water_bottle,milk_bottle,carton,can"
```
Comment (PhaniShekhar) [inline cli/foundation-models/system/inference/zero-shot-image-classification/prepare_data.py:11]: Nit: Add spaces after ',' ?

File: sdk/python/foundation-models/system/inference/zero-shot-image-classification/zero-shot-image-classification-online-endpoint.ipynb [hunk 4]
```diff
@@ -0,0 +1,357 @@
+{
+ "cells": [
+  {
+   "attachments": {},
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## Zero-Shot Image Classification Inference using Online Endpoints\n",
+    "\n",
+    "This sample shows how deploy `zero-shot-image-classification` type models to an online endpoint for inference.\n",
```
Comment (PhaniShekhar) [inline sdk/python/foundation-models/system/inference/zero-shot-image-classification/zero-shot-image-classification-online-endpoint.ipynb:10]: Nit: how to deploy

File: sdk/python/foundation-models/system/inference/zero-shot-image-classification/zero-shot-image-classification-online-endpoint.ipynb [hunk 5]
```diff
@@ -0,0 +1,357 @@
...
+    "    print(ex)\n",
+    "    # Enter details of your AML workspace\n",
+    "    subscription_id = \"<SUBSCRIPTION_ID>\"\n",
+    "    resource_group = \"<RESOURCE_GROUP>\"\n",
+    "    workspace_name = \"<AML_WORKSPACE_NAME>\"\n",
+    "workspace_ml_client = MLClient(\n",
+    "    credential, subscription_id, resource_group, workspace_name\n",
+    ")\n",
+    "\n",
+    "# The models, fine tuning pipelines and environments are available in the AzureML system registry, \"azureml\"\n",
```
Comment (PhaniShekhar) [inline sdk/python/foundation-models/system/inference/zero-shot-image-classification/zero-shot-image-classification-online-endpoint.ipynb:79]: Minor: fine tuning pipelines and environments are not relevant to our case, right ? Would it make sense to remove them in this comment.

File: sdk/python/foundation-models/system/inference/zero-shot-image-classification/zero-shot-image-classification-online-endpoint.ipynb [hunk 6]
```diff
@@ -0,0 +1,357 @@
...
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "import base64\n",
+    "import json\n",
+    "\n",
+    "sample_image = os.path.join(dataset_dir, \"milk_bottle\", \"99.jpg\")\n",
+    "labels = \"water_bottle,milk_bottle,carton,can\"\n",
```
Comment (PhaniShekhar) [inline sdk/python/foundation-models/system/inference/zero-shot-image-classification/zero-shot-image-classification-online-endpoint.ipynb:278]: Nit: Use spaces after ","

File: sdk/python/foundation-models/system/inference/zero-shot-image-classification/zero-shot-image-classification-online-endpoint.ipynb [hunk 7]
```diff
@@ -0,0 +1,357 @@
...
+    "    with open(image_path, \"rb\") as f:\n",
+    "        return f.read()\n",
+    "\n",
+    "\n",
+    "request_json = {\n",
+    "    \"input_data\": {\n",
+    "        \"columns\": [\"image\", \"text\"],\n",
+    "        \"index\": [0],\n",
+    "        \"data\": [\n",
+    "            [base64.encodebytes(read_image(sample_image)).decode(\"utf-8\"), labels]\n",
```
Comment (PhaniShekhar) [inline sdk/python/foundation-models/system/inference/zero-shot-image-classification/zero-shot-image-classification-online-endpoint.ipynb:291]: Should we add an example for invoking endpoint on multiple images at once ?

**Review by PhaniShekhar — CHANGES_REQUESTED** [review]

File: sdk/python/foundation-models/system/inference/zero-shot-image-classification/zero-shot-image-classification-online-endpoint.ipynb (hunk 7 above)
Comment (bhimar) [reply sdk/python/foundation-models/system/inference/zero-shot-image-classification/zero-shot-image-classification-online-endpoint.ipynb]: I think it is a good idea. I added a second image and comments to explain the request format

File: cli/foundation-models/system/inference/zero-shot-image-classification/prepare_data.py [hunk 8]
```diff
@@ -0,0 +1,111 @@
...
+        return f.read()
+
+
+def prepare_data_for_online_inference(dataset_dir: str) -> None:
+    """Prepare request json for online inference.
+
+    :param dataset_dir: dataset directory
+    :type dataset_dir: str
+    """
+    sample_image = os.path.join(dataset_dir, "milk_bottle", "99.jpg")
```
Comment (PhaniShekhar) [inline cli/foundation-models/system/inference/zero-shot-image-classification/prepare_data.py:69]: Should we update this also to use two images ?

File: sdk/python/foundation-models/system/inference/zero-shot-image-classification/zero-shot-image-classification-online-endpoint.ipynb (hunk 7 above)
Comment (PhaniShekhar) [reply sdk/python/foundation-models/system/inference/zero-shot-image-classification/zero-shot-image-classification-online-endpoint.ipynb]: I excluded the text column in the model card. Should we keep it consistent in the notebooks and examples: https://ml.azure.com/registries/azmlft-dev-registry01/models/openai-clip-vit-base-patch32/version/3?tid=72f988bf-86f1-41af-91ab-2d7cd011db47#overview

Comment (bhimar) [reply sdk/python/foundation-models/system/inference/zero-shot-image-classification/zero-shot-image-classification-online-endpoint.ipynb]: Follow up...turns out this doesn't fail!

**Review by PhaniShekhar — APPROVED** [review]

File: sdk/python/foundation-models/system/inference/zero-shot-image-classification/zero-shot-image-classification-online-endpoint.ipynb [hunk 9]
```diff
@@ -0,0 +1,364 @@
...
+    ")\n",
+    "\n",
+    "# The models are available in the AzureML system registry, \"azureml\"\n",
+    "registry_ml_client = MLClient(\n",
+    "    credential,\n",
+    "    subscription_id,\n",
+    "    resource_group,\n",
+    "    registry_name=\"azureml\",\n",
+    ")\n",
+    "# Generating a unique timestamp that can be used for names and versions that need to be unique\n",
```
Comment (rdondera-microsoft) [inline sdk/python/foundation-models/system/inference/zero-shot-image-classification/zero-shot-image-classification-online-endpoint.ipynb:86]: Hm, do we need the timestamp value created here?

Comment (bhimar) [reply sdk/python/foundation-models/system/inference/zero-shot-image-classification/zero-shot-image-classification-online-endpoint.ipynb:86]: For consistency with our other notebooks I'll keep it

**Pull request merged**